### PR TITLE
Hermez client

### PR DESCRIPTION
**Pull request by nitinmittal23**
This PR adds `HermezClient` implementation which covers the bridge functionality of hermez chain.

### CODE REVIEW

File: src/hermez/erc20.ts [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+    }
+
+    /**
+     * is Approval needed to bridge tokens to other chains
+     *
+     * @returns
+     * @memberof ERC20
+     */
+    isApprovalNeeded() {
+        if (this.contractParam.address === ADDRESS_ZERO) {
```
Comment (ujjwalgupta94) [inline src/hermez/erc20.ts:68]: wrap this logic into a method `isEtherToken` and then use it at multiple place

File: src/hermez/erc20.ts [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+    deposit(amount: TYPE_AMOUNT, userAddress: string, option: IBridgeTransactionOption = {}) {
+        this.checkForRoot("deposit");
+        const permitData = option.permitData || '0x';
+
+        const amountInABI = this.client.parent.encodeParameters(
+            [Converter.toHex(amount)],
+            ['uint256'],
+        );
+
+        if (this.contractParam.address === ADDRESS_ZERO) {
```
Comment (ujjwalgupta94) [inline src/hermez/erc20.ts:157]: same, use isEtherToken here

File: src/utils/set_proof_api_url.ts [hunk 3]
```diff
@@ -7,4 +7,14 @@ export const setProofApi = (url: string) => {
     }
     url += 'api/v1/';
     service.network = new NetworkService(url);
-};
\ No newline at end of file
+};
+
+export const setHermezProofApi = (url: string, override = true) => {
```
Comment (ujjwalgupta94) [inline src/utils/set_proof_api_url.ts:12]: what is need of additional param, override ?

I think  if user calls it which means he wants to override .

File: src/abstracts/base_web3_client.ts [hunk 4]
```diff
@@ -24,6 +25,11 @@ export abstract class BaseWeb3Client {
 
     abstract getBlock(blockHashOrBlockNumber): Promise<IBlock>;
     abstract getBlockWithTransaction(blockHashOrBlockNumber): Promise<IBlockWithTransaction>;
+    abstract hexToNumber(value: any): number;
+    abstract hexToNumberString(value: any): string;
```
[the remark this answers is not in the document]
Comment (nitinmittal23) [reply src/abstracts/base_web3_client.ts:29]: its just one will give number and other will give string. as sometimes large number do not work in hexToNumber so convert it to string

File: src/constant.ts [hunk 5]
```diff
@@ -1 +1,11 @@
 export const MAX_AMOUNT = '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff';
+export const ADDRESS_ZERO = '0x0000000000000000000000000000000000000000'
+export const DAI_PERMIT_TYPEHASH = "0xea2aa0a1be11a07ed86d755c93467f4f82362b452371d1ba94d1715123511acb";
```
[the remark this answers is not in the document]
Comment (nitinmittal23) [reply src/constant.ts:3]: actually domains are common but will be different for some other tokens. these are standard which we can add for now.

File: src/utils/set_proof_api_url.ts (hunk 3 above)
Comment (nitinmittal23) [reply src/utils/set_proof_api_url.ts:12]: so in hermez client init function we set this hermezProofApi by default and we don't want that part of the code to over write the user API that is why in that function we are using the value 'false'

File: src/abstracts/base_web3_client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ujjwalgupta94) [reply src/abstracts/base_web3_client.ts:29]: ok make sense, but the name `hexToNumberString` is confusing. Can we make it `hexToString` ?

File: src/abstracts/base_web3_client.ts [hunk 6]
```diff
@@ -36,6 +42,17 @@ export abstract class BaseWeb3Client {
         });
     }
 
+    getAccounts_() {
```
[the remark this answers is not in the document]
Comment (ujjwalgupta94) [reply src/abstracts/base_web3_client.ts:45]: ok works, thanks

File: src/constant.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ujjwalgupta94) [reply src/constant.ts:3]: That is the reason,  i didn't wanted to add this into maticjs.

File: src/abstracts/base_web3_client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (nitinmittal23) [reply src/abstracts/base_web3_client.ts:29]: Number string makes much sense and we are converting it to a number first. else hex to string means we are converting only hex to a string..

**Pull request merged**